### PR TITLE
New FastSim Hit Reconstruction Emulation

### PR DESCRIPTION
**Pull request by matt-komm**
A new producer to emulate hit reconstruction for tracking in FastSim is introduced. It features
- simple plugin-based interface to produce reconstructed hits
- plugins can be restricted to certain DetIds
- hits can be merged if SimHits are too close

For comparisons, the old producer can still be used but will be removed in the future. Currently, merging of hits is still deactivated. This will be changed in a later PR.

Related presentations from the 2nd FastSim days (https://indico.cern.ch/event/473392/):
- https://indico.cern.ch/event/473392/contributions/1994037/attachments/1229086/1800822/2016_02_16.pdf
- https://indico.cern.ch/event/473392/contributions/1994039/attachments/1229120/1801774/fastsim_alicecocoros.pdf

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild , please test

File: Configuration/Generator/python/SingleMuPt100_pythia8_cfi.py [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import FWCore.ParameterSet.Config as cms
 generator = cms.EDFilter("Pythia8PtGun",
                          PGunParameters = cms.PSet(
-        MaxPt = cms.double(100.01),
-        MinPt = cms.double(99.99),
+        MaxPt = cms.double(100.0),
+        MinPt = cms.double(0.),
```
Comment (lveldere) [inline Configuration/Generator/python/SingleMuPt100_pythia8_cfi.py]: undo the changes to this file

File: FastSimulation/Configuration/python/Reconstruction_BefMix_cff.py [hunk 2]
```diff
@@ -13,7 +13,12 @@
 from RecoVertex.BeamSpotProducer.BeamSpot_cff import offlineBeamSpot
 
 # and of course it needs tracker hits
-from FastSimulation.TrackingRecHitProducer.SiTrackerGaussianSmearingRecHitConverter_cfi import siTrackerGaussianSmearingRecHits
+
+#uses the old RecHit producer
+#from FastSimulation.TrackingRecHitProducer.SiTrackerGaussianSmearingRecHitConverter_cfi import siTrackerGaussianSmearingRecHits
+#uses the new RecHit producer
+from FastSimulation.TrackingRecHitProducer.NewRecHitConverter_NoMerge_cfi import trackingRecHitProducerNoMerge as siTrackerGaussianSmearingRecHits
```
Comment (lveldere) [inline FastSimulation/Configuration/python/Reconstruction_BefMix_cff.py]: use proper file names and module names
(no *New*RecHitConverter..., get rid of the name siTrackerGaussianSmearingRecHits)

File: FastSimulation/TrackingRecHitProducer/python/SiTrackerGaussianSmearingRecHitConverter_cfi.py [hunk 3]
```diff
@@ -1,8 +1,5 @@
 import FWCore.ParameterSet.Config as cms
 
-
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/python/SiTrackerGaussianSmearingRecHitConverter_cfi.py]: get rid of this file and all remnants of
SiTrackerGaussianSmearingRecHitConverter

File: FastSimulation/TrackingRecHitProducer/python/NewRecHitConverter_NoMerge_cfi.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+trackingRecHitProducerNoMerge.plugins.append(
+    cms.PSet(
+        name = cms.string("pixelBarrelSmearer"),
+        type=cms.string("PixelBarrelTemplateSmearerPlugin"),
+        NewPixelBarrelResolutionFile1 = cms.string('FastSimulation/TrackingRecHitProducer/data/NewPixelResolutionBarrel38T.root'),
+        NewPixelBarrelResolutionFile2 = cms.string('FastSimulation/TrackingRecHitProducer/data/NewPixelResolutionBarrelEdge38T.root'),
+        NewPixelBarrelResolutionFile3 = cms.string('FastSimulation/TrackingRecHitProducer/data/PixelBarrelResolution2014.root'),
+        NewPixelForwardResolutionFile = cms.string('FastSimulation/TrackingRecHitProducer/data/NewPixelResolutionForward38T.root'),
+        NewPixelForwardResolutionFile2 = cms.string('FastSimulation/TrackingRecHitProducer/data/PixelForwardResolution2014.root'),
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/python/NewRecHitConverter_NoMerge_cfi.py]: please introduce meaningful parameter names and meaningful data file names

File: FastSimulation/TrackingRecHitProducer/python/NewRecHitConverter_NoMerge_cfi.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/python/NewRecHitConverter_NoMerge_cfi.py]: give this file a decent name
(no *New*RecHitConverter...)

**Comment by lveldere** [comment]
@matt-komm 

please move all data files to the following repository:
https://github.com/cms-data/FastSimulation-TrackingRecHitProducer
(make a pr, or open an issue)
remove all files from FastSimulation/TrackingRecHitProducer/data/

About the 2 files already available in this repository:
PixelBarrelResolution2014.root
PixelForwardResolution2014.root
Only replace them if you are sure they need to be replaced.

File: FastSimulation/TrackingRecHitProducer/test/RunTrackingReco.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/test/RunTrackingReco.py]: remove this file, or replace with a cmsDriver command

File: FastSimulation/TrackingRecHitProducer/test/RunTrackingReco_real.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/test/RunTrackingReco_real.py]: remove this file, or replace with a cmsDriver command

File: FastSimulation/TrackingRecHitProducer/test/SimulateSimHits.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
+# Auto generated configuration file
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/test/SimulateSimHits.py]: remove this file, or replace with a cmsDriver command

File: FastSimulation/TrackingRecHitProducer/interface/PixelTemplateSmearerBase.h [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        std::map<unsigned int, const SimpleHistogramGenerator*> theXHistos;
+        std::map<unsigned int, const SimpleHistogramGenerator*> theYHistos;
+
+        TFile* thePixelResolutionFile1;
+        std::string thePixelResolutionFileName1;
+        TFile* thePixelResolutionFile2;
+        std::string thePixelResolutionFileName2;
+        TFile* thePixelResolutionFile3;
+        std::string thePixelResolutionFileName3;
+        TFile* probfile;
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/interface/PixelTemplateSmearerBase.h]: please introduce meaningful names for the resolution files

File: FastSimulation/TrackingRecHitProducer/plugins/PixelBarrelTemplateSmearerPlugin.cc [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    PixelTemplateSmearerBase(name,config,consumesCollector)
+{
+    isForward = false;
+    thePixelResolutionFileName1 = config.getParameter<string>( "NewPixelBarrelResolutionFile1" );
+    thePixelResolutionFile1 = new TFile( edm::FileInPath( thePixelResolutionFileName1 ).fullPath().c_str()  ,"READ");
+    thePixelResolutionFileName2 =  config.getParameter<string>( "NewPixelBarrelResolutionFile2" );
+    thePixelResolutionFile2 = new TFile( edm::FileInPath( thePixelResolutionFileName2 ).fullPath().c_str()  ,"READ");
+    thePixelResolutionFileName3 = config.getParameter<string>( "NewPixelBarrelResolutionFile3" );
+    thePixelResolutionFile3 = new TFile( edm::FileInPath( thePixelResolutionFileName3 ).fullPath().c_str()  ,"READ");
+
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/plugins/PixelBarrelTemplateSmearerPlugin.cc]: please introduce meaningful parameter names

File: FastSimulation/TrackingRecHitProducer/interface/PixelTemplateSmearerBase.h [hunk 11]
```diff
@@ -0,0 +1,129 @@
+#ifndef FastSimulation_TrackingRecHitProducer_PixelTemplateSmearerBase_h
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/interface/PixelTemplateSmearerBase.h]: Base class is not generic enough.
Why is a base class needed?

File: FastSimulation/TrackingRecHitProducer/plugins/PixelBarrelTemplateSmearerPlugin.cc [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/plugins/PixelBarrelTemplateSmearerPlugin.cc]: I doubt there is need for 2 implementations
PixelBarrelTemplateSmearerPlugin and
PixelForwardTemplateSmearerPlugin

I think that just one plugin will do and it can enherrit directly from trackingRecHitAlgorithm.
I.e. PixelTemplateSmearer should, after some small changes, become the plugin.

File: FastSimulation/TrackingRecHitProducer/src/PixelTemplateSmearerBase.cc [hunk 13]
```diff
@@ -0,0 +1,1001 @@
...
+    // protection against out-of-range (undeflows and overflows)
+    if (cotalphaHistBin < 1) cotalphaHistBin = 1; 
+    if (cotbetaHistBin  < 1) cotbetaHistBin  = 1; 
+    if (cotalphaHistBin > (int)rescotAlpha_binN) cotalphaHistBin = (int)rescotAlpha_binN; 
+    if (cotbetaHistBin  > (int)rescotBeta_binN) cotbetaHistBin  = (int)rescotBeta_binN; 
+    //
+    unsigned int theXHistN;
+    unsigned int theYHistN;
+
+    if (!isForward)
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/src/PixelTemplateSmearerBase.cc]: The generation of the key to find backt the right histogram in the template map is implemented twice.
Once here and once in the initialisation of
PixelForwardTemplateSmearerPlugin and PixelBarrelTemplateSmearerPlugin.
One should write a function, e.g. something like
getSmearingTemplateKey(
    cotalphaHistBin,
    cotbetaHistBin,
    isForward,
    edge,
    singlex,
    hitbigx,
    hasBigPixelInX,
    singley,
    hitbigy,
    hasBigPixelInY)

File: FastSimulation/TrackingRecHitProducer/src/PixelTemplateSmearerBase.cc [hunk 14]
```diff
@@ -0,0 +1,1001 @@
...
+}
+
+
+FastSingleTrackerRecHit PixelTemplateSmearerBase::
+smearMergeGroup(
+    MergeGroup* mg,
+    const PixelGeomDetUnit * detUnit,
+    const double boundX, const double boundY,
+    RandomEngineAndDistribution const * random
+) const
```
Comment (matt-komm) [inline FastSimulation/TrackingRecHitProducer/src/PixelTemplateSmearerBase.cc]: this method copies some code from the `smearHit` method. Can it be moved into an extra method?

**Comment by lveldere** [comment]
@matt-komm 

Could you add a file 
FastSimulation/TrackingRecHitProducer/README.md 
to guide people a bit through the package?

File: FastSimulation/TrackingRecHitProducer/plugins/SiTrackerGaussianSmearingRecHitConverter.cc [hunk 15]
```diff
@@ -41,6 +41,8 @@
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/plugins/SiTrackerGaussianSmearingRecHitConverter.cc]: remove this file and all other obsolete files from package

File: FastSimulation/TrackingRecHitProducer/plugins/TrackingRecHitProducer.cc [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+
+        TrackingRecHitAlgorithm* recHitAlgorithm = TrackingRecHitAlgorithmFactory::get()->tryToCreate(pluginType,pluginName,pluginConfig,consumeCollector);
+        if (recHitAlgorithm)
+        {
+            edm::LogInfo("TrackingRecHitProducer: ")<< "adding plugin type '"<<pluginType<<"' as '"<<pluginName<<"'"<<std::endl;
+            _recHitAlgorithms.push_back(recHitAlgorithm);
+        }
+        else
+        {
+            edm::LogWarning("TrackingRecHitAlgorithm plugin not found: ") << "plugin type = "<<pluginType<<"\nconfiguration=\n"<<pluginConfig.dump();
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/plugins/TrackingRecHitProducer.cc]: throw an error instead

File: FastSimulation/TrackingRecHitProducer/plugins/TrackingRecHitProducer.h [hunk 17]
```diff
@@ -0,0 +1,71 @@
+#ifndef FastSimulation_TrackingRecHitProducer_TrackingRecHitProducer_h
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/plugins/TrackingRecHitProducer.h]: For plugins I prefer to have no .h file.

File: FastSimulation/TrackingRecHitProducer/plugins/TrackingRecHitStripGSSmearingPlugin.cc [hunk 18]
```diff
@@ -0,0 +1,118 @@
+#include "FastSimulation/TrackingRecHitProducer/interface/TrackingRecHitAlgorithm.h"
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/plugins/TrackingRecHitStripGSSmearingPlugin.cc]: class name:
or ...GSPlugin
or ...GaussianSmearingPlugin

**Comment by lveldere** [comment]
@matt-komm 

Thanks a lot!
I'm done with the first iteration of reviewing.
Let's talk over skype once you had a look.

**Comment by matt-komm** [comment]
We finished the code revision!

**Comment by lveldere** [comment]
please test

**Comment by matt-komm** [comment]
tests fail because https://github.com/cms-data/FastSimulation-TrackingRecHitProducer/pull/2 has not be merged yet

File: FastSimulation/TrackingRecHitProducer/plugins/TrackingRecHitProducer.cc [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+    edm::ConsumesCollector consumeCollector = consumesCollector();
+    const std::vector<edm::ParameterSet>& pluginConfigs = config.getParameter<std::vector<edm::ParameterSet>>("plugins");
+
+    for (unsigned int iplugin = 0; iplugin<pluginConfigs.size(); ++iplugin)
+    {
+        const edm::ParameterSet& pluginConfig = pluginConfigs[iplugin];
+        const std::string pluginType = pluginConfig.getParameter<std::string>("type");
+        const std::string pluginName = pluginConfig.getParameter<std::string>("name");
+
+        TrackingRecHitAlgorithm* recHitAlgorithm = TrackingRecHitAlgorithmFactory::get()->tryToCreate(pluginType,pluginName,pluginConfig,consumeCollector);
```
Comment (lveldere) [inline FastSimulation/TrackingRecHitProducer/plugins/TrackingRecHitProducer.cc]: @matt-komm 

where is the deletion of recHitAlgorithm taken care of?

Comment (matt-komm) [inline FastSimulation/TrackingRecHitProducer/plugins/TrackingRecHitProducer.cc]: just added proper deletion of these plugins - however, this would not have lead to an increasing memleak

**Comment by smuzaffar** [comment]
cms-sw/cmsdist#2347 is now merged and new data files shall be available in 23h IB.

**Comment by lveldere** [comment]
please test

**Comment by lveldere** [comment]
+1

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
+1

**Comment by davidlange6** [comment]
@matt-komm - while reviewing the rest - we need to move all the root files to the external data areas. 
@smuzaffar @degano @iahmad-khan

**Comment by smuzaffar** [comment]
@iahmad-khan , can you please take care of it?

**Comment by iahmad-khan** [comment]
@davidlange6 the root files in this PR has been deleted already and they are present in cms-data /FastSimulation-TrackingRecHitProducer

**Comment by davidlange6** [comment]
@matt-komm , @iahmad-khan - the root files are still in the PR itself - so if they are now in externals, we need @matt-komm to remove them from the PR itself. Thanks!

**Comment by matt-komm** [comment]
@davidlange6 I cannot see a data directory with files inside this PR anymore (https://github.com/matt-komm/cmssw/tree/hitmerging-master/FastSimulation/TrackingRecHitProducer); I also checked it out in a fresh IB and these old data files are also not there anymore - can you let me know how you spotted them after this commit https://github.com/cms-sw/cmssw/pull/14843/commits/c1943d81f97db98461e469bcc3d5350fb8d06a06 ?

**Comment by lveldere** [comment]
@davidlange6 

Hi David

With respect to root files,
This pr only *removes* root files.
(check the "files chaged" for this pr)

Cheers

Lukas

**Comment by davidlange6** [comment]
Whoops, my mistake!
